### DISCUSSION TITLE
Is it possible to generate signatures with IntPtr rather than structs?

### DISCUSSION THREAD

**Discussion by netcorefan1** [microsoft/CsWin32#1090] [Q&A] [answered]
Hello,
I'm converting my old code into CsWin32, but I'm having troubles with `SetWinEventHook` and `UnhookWinEvent` which generates the following signatures:
```
public static extern HWINEVENTHOOK SetWinEventHook(uint eventMin, uint eventMax, HMODULE hmodWinEventProc, WINEVENTPROC pfnWinEventProc, uint idProcess, uint idThread, uint dwFlags);
public static extern BOOL UnhookWinEvent(HWINEVENTHOOK hWinEventHook);
```

I use my own `SafeHandle` which inherits from `SafeHandleZeroOrMinusOneIsInvalid`.  It's `SafeHandle.SetHandle()` requires a parameter of type `IntPtr`. Problem is when it comes the time to call `ReleaseHandle()` override. I can't call `UnhookWinEvent` from there  because it only accepts `HWINEVENTHOOK`.
This means that I'm forced to manually Dllimport the required signature which basically its against the main purpose of CsWin32.

I had the same problem with other API calls too where `IntPtr` type is wrapped around custom structs. Most of times I noticed that they are the same structures (same operators, equality checks, functions etc), the only change is the name which unfortunately is enough to break interoperability with other functions. Is there a particular reason for wrapping this type around so many identical structs?

What is the recommended way to handle such cases? Any way to preserve `IntPtr` for specific api calls?

**Comment by jnm2** [comment, marked as answer]
> I can't call `UnhookWinEvent` from there because it only accepts `HWINEVENTHOOK`.

Why can't you cast the IntPtr to `HWINEVENTHOOK` to make the call?

**Reply by netcorefan1** [reply]
Many thanks @jnm2 ! This is what I have done. To make this work, I had to change the code and add an additional generic parameter for `HWINEVENTHOOK`.
Now:
`.AsSafe(ptr => UnhookWinEvent((HWINEVENTHOOK)ptr))`
Before:
`.AsSafe(UnhookWinEvent)`
Not good as it was before, but at least I have not been forced to manually add a dllimport declaration.
